### ISSUE TITLE
How to expose port 22 on a pod to the outside world?

### ISSUE THREAD

**Issue by esnible** [istio/istio#32169]
I am struggling to expose port 22 using an Istio Gateway.  Are there any examples?

Maybe I have made a rookie mistake?  `istioctl analyze` reports nothing wrong, so if there is anything wrong in my config I'd like to tackle it with an analyzer.

First I deploy Istio with an extra gateway listening on port 22.  Initially I didn't want to do this, but I want to use `scp` to talk to my pod, and I have no DNS name for my pod, so I was concerned I wouldn't be able to use the existing gateway.  I re-installed Istio 1.9.2 using the second IOP in https://github.com/istio/istio/issues/32166

First I deploy a pod that listens on port 22: `kubectl run panubo-sshd --image docker.io/panubo/sshd:1.3.0 --port=22 --env="SSH_USERS=user:1000:1000" --env="SSH_ENABLE_PASSWORD_AUTH=true"`

Next I tried to expose the pod:

``` bash
cat > /tmp/expose-22.yaml <<EOF
apiVersion: v1
kind: Service
metadata:
  name: panubo
  labels:
    app: panubo-sshd
    service: panubo-sshd
spec:
  ports:
  - name: tls
    port: 22
  selector:
    run: panubo-sshd
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
  namespace: default
spec:
  selector:
    ssh-port: "22"
  servers:
  - hosts:
    - panubo.test
    port:
      number: 22
      name: ssh_server
      protocol: TLS
    tls:
      mode: PASSTHROUGH
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: panubo
spec:
  gateways:
  - mygateway
  hosts:
  - panubo.test
  tcp:
  - match:
    - port: 22
    route:
    - destination:
        host: panubo
        port:
          number: 22
EOF
kubectl apply -f /tmp/expose-22.yaml
```

(Everything in the above is on port 22.  I thought it would be cool to have the ingressgateway listen on port 3022, and use `scp -P 3022`, but I was confused by the output of `istioctl analyze` when I had a port/targetPort.  I was also confused by the statement at https://istio.io/latest/docs/reference/config/networking/gateway/#Port that the `targetPort` on a Gateway is "Applicable only when used with ServiceEntries." so I decided to use 22 throughout).

`kubectl -n istio-system get svc ssh-gateway` shows the port as `22:30731` so I was really hoping that the pod would be listening on one of those ports but it isn't:

``` bash
kubectl -n istio-system exec deployment/ssh-gateway -- netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      16/envoy            
tcp        0      0 0.0.0.0:15090           0.0.0.0:*               LISTEN      16/envoy            
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      16/envoy            
tcp        0      0 0.0.0.0:15021           0.0.0.0:*               LISTEN      16/envoy            
tcp6       0      0 :::15020                :::*                    LISTEN      1/pilot-agent   
```

Any idea what I am doing wrong?

**Comment by howardjohn** [comment]
Does the gateway pod you deployed have labels
```
  selector:
    ssh-port: "22"
```

?

**Comment by sridharlreddy** [comment]
Yes simple mistake - change the match in Virtual Service from tcp to tls or use tcp in Gateway.

**Comment by esnible** [comment]
> Yes simple mistake - change the match in Virtual Service from tcp to tls or use tcp in Gateway.

@sridharlreddy Were you able to get the IngressGateway to listen on port 22, or 3022?  I changed the VirtualService to TLS and it did not work for me.

``` bash
cat <<EOF | kubectl apply -f -
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: panubo
spec:
  gateways:
  - mygateway
  hosts:
  - "*"
  tls:
  - match:
    - port: 22
      sniHosts: ["*"]
    route:
    - destination:
        host: panubo
        port:
          number: 22
EOF
```

As before, `istioctl analyze` reported nothing wrong.  I'd like to get this working, then fix `analyze` to suggest the solution.

**Comment by esnible** [comment]
> Does the gateway pod you deployed have labels
> 
> ```
>   selector:
>     ssh-port: "22"
> ```

```bash
$ kubectl -n istio-system get pod ssh-gateway-69586967d5-28nfk -o jsonpath="{.metadata.labels}"
{"app":"istio-ingressgateway","chart":"gateways","heritage":"Tiller","install.operator.istio.io/owning-resource":"unknown","istio":"ingressgateway","istio.io/rev":"default","operator.istio.io/component":"IngressGateways","pod-template-hash":"69586967d5","release":"istio","service.istio.io/canonical-name":"ssh-gateway","service.istio.io/canonical-revision":"latest","sidecar.istio.io/inject":"false","ssh-port":"22"}
```

**Comment by howardjohn** [comment]
are you hitting `skipping privileged gateway port 22 for node istio-ingressgateway-589754c4dd-j9kl6.istio-system as it is an unprivileged pod` in istiod logs?

**Comment by esnible** [comment]
> unprivileged

No.  `kubectl -n istio-system logs deployment/istiod | grep -i unprivileged` has no output.  The only thing skipped in the logs is cert rotation.

**Comment by sridharlreddy** [comment]
> > Yes simple mistake - change the match in Virtual Service from tcp to tls or use tcp in Gateway.
> 
> @sridharlreddy Were you able to get the IngressGateway to listen on port 22, or 3022? I changed the VirtualService to TLS and it did not work for me.
> 
> ```shell
> cat <<EOF | kubectl apply -f -
> apiVersion: networking.istio.io/v1alpha3
> kind: VirtualService
> metadata:
>   name: panubo
> spec:
>   gateways:
>   - mygateway
>   hosts:
>   - "*"
>   tls:
>   - match:
>     - port: 22
>       sniHosts: ["*"]
>     route:
>     - destination:
>         host: panubo
>         port:
>           number: 22
> EOF
> ```
> 
> As before, `istioctl analyze` reported nothing wrong. I'd like to get this working, then fix `analyze` to suggest the solution.

Yes, both 22 with privileged and 3022 unprivileged works. istioctl analyze did not report any error.
```

k -n istio-system exec -ti ssh-gateway-788dc59767-l4swd -- netstat -nlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 0.0.0.0:15021           0.0.0.0:*               LISTEN      14/envoy            
tcp        0      0 0.0.0.0:3022            0.0.0.0:*               LISTEN      14/envoy            
tcp        0      0 0.0.0.0:15090           0.0.0.0:*               LISTEN      14/envoy            
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      14/envoy            
tcp6       0      0 :::15020                :::*                    LISTEN      1/pilot-agent     

```

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    app: ssh
  servers:
  - hosts:
    - '*'
    port:
      number: 22
      name: tls
    #  protocol:  TCP
      protocol:  TLS
    tls:
      mode: PASSTHROUGH
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: panubo
spec:
  gateways:
  - mygateway
  hosts:
  - '*'
  #tcp:
  tls:
  - match:
    - port: 22
      sniHosts:
        - '*'
    route:
    - destination:
        host: ssh
        port:
          number: 22
```

ssh was failing with banner error for panubo image, I used linux-server/openssh-server image and it worked.

**Comment by esnible** [comment]
> ```
> k -n istio-system exec -ti ssh-gateway-788dc59767-l4swd -- netstat -nlp
> Active Internet connections (only servers)
> Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
> tcp        0      0 0.0.0.0:15021           0.0.0.0:*               LISTEN      14/envoy            
> tcp        0      0 0.0.0.0:3022            0.0.0.0:*               LISTEN      14/envoy            
> tcp        0      0 0.0.0.0:15090           0.0.0.0:*               LISTEN      14/envoy            
> tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      14/envoy            
> tcp6       0      0 :::15020                :::*                    LISTEN      1/pilot-agent     
> ```

@sridharlreddy can you attach what I need to get a gateway listening on port 3022 like this?  I haven't been able to get it working.

I'm using

```
kubectl run open-sshd --image docker.io/linuxserver/openssh-server --port=22 --env="SSH_USERS=user:1000:1000" --env="SSH_ENABLE_PASSWORD_AUTH=true"

``` bash
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: ssh
  labels:
    app: open-sshd
    service: open-sshd
spec:
  ports:
  - name: tls
    port: 22
  selector:
    run: open-sshd
EOF

cat <<EOF | kubectl apply -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    app: istio-ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      number: 3022
      targetPort: 22
      name: tls
      protocol:  TLS
    tls:
      mode: PASSTHROUGH
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: panubo
spec:
  gateways:
  - mygateway
  hosts:
  - '*'
  tls:
  - match:
    - port: 22
      sniHosts:
        - '*'
    route:
    - destination:
        host: ssh
        port:
          number: 22
EOF
```

I'm adding 3022 to the istio-ingressgateway by installing using

``` bash
cat > /tmp/ingress-3022.yaml <<EOF
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  components:
    ingressGateways:
    - name: istio-ingressgateway
      k8s:
        service:
          ports:
          - port: 15021
            targetPort: 15021
            name: status-port
          - port: 80
            targetPort: 8080
            name: http2
          - port: 443
            targetPort: 8443
            name: https
          - port: 31400
            targetPort: 31400
            name: tcp
          - port: 15443
            targetPort: 15443
            name: tls
          - port: 3022
            name: tls-ssh
EOF
~/go/src/istio.io/istio/out/darwin_amd64/istioctl install  -f /tmp/ingress-3022.yaml
```

**Comment by esnible** [comment]
Related: https://github.com/istio/istio/issues/28259

**Comment by sridharlreddy** [comment]
@esnible 


Change istio operator input as below.
```
      k8s:
        service:
          ports:
          - port: 22
             targetPort: 3022
            name: tls-ssh
```

Use just port 22 in Virtual Service, it works. **3022** is defined only here, rest everywhere it is 22.

**Comment by esnible** [comment]
@sridharlreddy Thank you for taking the time to help me.  I was able to get it working.

After incorporating sridharlreddy's suggestions, the IngressGateway listens on port 22 (the service) and port 3022 (the pod).

I still can't `ssh`.  The next problem, which I believe I am now past, is that I misunderstood sridharlreddy's suggestion to "Virtual Service from tcp to tls or use tcp in Gateway."  I changed the VS to use TLS.  That was exactly the wrong thing to do because `ssh` is an old program that pre-dates TLS and doesn't use it.  The gateway rejected my contact.  If Envoy logging was lowered to debug it logged "TLS error: 268435703:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER"

At that point I remembered that SSH doesn't use TLS so I went to TCP.  This seems to be working from an Istio point-of-view.  I still can't get in, but probably that is a misconfiguration of my `docker.io/linuxserver/openssh-server` pod.  Istio's ingress-gatway logs "wasm log stats_outbound tcp_stats_outbound ..." which makes me think it is forwarding.

I haven't yet come up with an analyzer design that to warn others of my problem.

For anyone following along, I am still using the IOP from https://github.com/istio/istio/issues/32169#issuecomment-826849759 to install Istio, and I'm using this Istio configuration:

``` bash
cat <<EOF | kubectl apply -f -
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: ssh-gateway
spec:
  selector:
    app: istio-ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      number: 22
      targetPort: 22
      name: tcp
      protocol: TCP
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: ssh-vs
spec:
  gateways:
  - ssh-gateway
  hosts:
  - '*'
  tcp:
  - match:
    - port: 22
    route:
    - destination:
        host: ssh
        port:
          number: 22
EOF
```

Because `ssh` doesn't support SNI a unique port will be needed for each ssh server that receives traffic.
